### DISCUSSION TITLE
Dependency injection at user session level

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [fastapi/fastapi#12259] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
n/a
```

### Description

A user makes multiple requests to a fastapi server during their session. The function processing the request(n+1) at the server will use information extracted from the request(n) payload. The request(n) payload is overwritten by the request(n+1) payload. And, so on.

Is this possible with fastapi?

### Operating System

Windows

### FastAPI Version

0.114.0

### Pydantic Version

2.7.3

### Python Version

3.11.4

**Comment by ceb10n** [comment]
I'm not sure if I understood it well, but what you want is: Access some info from a previous request, and then override it with new info?

You can try something with cookies, maybe?

```python
from typing import Annotated
from fastapi import Cookie, FastAPI, Response


app = FastAPI()


@app.get("/")
async def my_session_route(response: Response, my_session: Annotated[str | None, Cookie()] = None):
    next_val = "1" if my_session is None else int(my_session) + 1
    response.set_cookie(key="my_session", value=next_val)
    
    return {
        "ok": next_val
    }
```

**Reply by dineshbvadhia** [reply]
Problem understood well ie. access info from previous request and then overwrite with new info. Thank-you and will try this out.

**Reply by dineshbvadhia** [reply]
@ceb10n I may have stated the problem incorrectly bc by 'info' I meant a data object eg. a dictionary. Looking at https://www.starlette.io/responses/#set-cookie it says the value of a cookie must be a string value. 

I have edited the original question to clarify the problem. Thanks.

**Reply by sinisaos** [reply]
@dineshbvadhia Sorry if I missed your point, but you could try using [request.state](https://www.starlette.io/requests/#other-state) to get and set the payload data for each request. Something like this. 
```python
from fastapi import FastAPI, Request

app = FastAPI()

PAYLOAD = {"arg": 0}


@app.get("/")
async def my_session_route(request: Request):
    # update payload value for current request
    PAYLOAD["arg"] += 1
    # set next value to request state
    request.state.next_val = PAYLOAD
    print(f"Request number {request.state.next_val['arg']}")
    return {"data": request.state.next_val}
```

This is a simplified example that shows how data is changed on each request and stored in the request state. Hope that helps

**Reply by ceb10n** [reply]
> @ceb10n I may have stated the problem incorrectly bc by 'info' I meant a data object eg. a dictionary. Looking at https://www.starlette.io/responses/#set-cookie it says the value of a cookie must be a string value.
> 
> I have edited the original question to clarify the problem. Thanks.

You can still serialize your data and save it in your cookie.

Other option is to save your state in a cache like redis, or a database like postgres, mongodb, etc.

Saving the session in cookies, caches and databases won't require you to worry if you need more then one instance of your service.

If you use something like state, you won't be able to horizontal scale your application

**Reply by dineshbvadhia** [reply]
@ceb10n Ah, ok about serializing the payload data - it is already in json. 

As there will be multiple users, does fastapi keep a track of each user session cookie?

The application is designed to scale horizontally.

**Reply by ceb10n** [reply, marked as answer]
If you need to scale horizontally, you can't rely on the server.

In my view, the best way to keep a user session is using cache, like redis, memcached, dragonfly, etc... It's fast and server independent

I don't know what kind of application you are building, but it's a good thing to keep in mind is that REST is stateless. Since fastapi is popular in this area, maybe you want to keep track of your user's state in the client, not in the APIs. But it's not a rule and you should go with what is best for your use case ;)

**Reply by dineshbvadhia** [reply]
@ceb10n I'm building an API-based application server. Keeping user state in the client is not an option because only developers have access to the client. So, for my use-case the user state is tracked in the API.

Agree the optimal solution is a cache. In my case, how is a unique id obtained for each user session?

**Reply by ceb10n** [reply]
@dineshbvadhia I think it's better if you message me on linkedin or whatsapp so we can continue this chat 😄

But in general, what you can do is:

When your user authenticate, you will generate a JWT token or a cookie to keep track of who is, etc. This will have the identification of the user
Then, when the user make a request, you'll extract the identification from the jwt token or cookie, and use it as a key on your cache.

**Reply by dineshbvadhia** [reply]
@ceb10n Ok. Message sent. Closed discussion.
